pkg/clusteragent/externalmetrics/store.go

```diff
@@ -0,0 +1,181 @@
+		},
+	}
+}
+
+// RegisterObserver registers an observer that will be notified when changes happen in the store
+// Current implementation does not scale beyond a handful of observers.
+// Calls are made synchronously to each observer for each operation.
+// The store guarantees that any lock has been released before calling observers.
+func (ds *DatadogMetricsInternalStore) RegisterObserver(observer DatadogMetricInternalObserver) {
+	ds.observersLock.Lock()
+	defer ds.observersLock.Unlock()
+
+	addObserver := func(operationType storeOperation, observerFunc DatadogMetricInternalObserverFunc) {
+		if observerFunc != nil {
+			ds.observers[operationType] = append(ds.observers[operationType], observerFunc)
+		}
+	}
+
+	addObserver(setOperation, observer.SetFunc)
+	addObserver(deleteOperation, observer.DeleteFunc)
+}
+
+// Get returns `DatadogMetricInternal` for given id, returns nil if absent
+func (ds *DatadogMetricsInternalStore) Get(id string) *model.DatadogMetricInternal {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	res, ok := ds.store[id]
+	if !ok {
+		return nil
+	}
+
+	return &res
+}
+
+// GetAll returns a copy of all store values
+func (ds *DatadogMetricsInternalStore) GetAll() []model.DatadogMetricInternal {
+	return ds.GetFiltered(func(model.DatadogMetricInternal) bool { return true })
+}
+
+// GetAll returns a copy of all store values matched by the `filter` function
+func (ds *DatadogMetricsInternalStore) GetFiltered(filter func(model.DatadogMetricInternal) bool) []model.DatadogMetricInternal {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	datadogMetrics := make([]model.DatadogMetricInternal, 0, len(ds.store))
+	for _, datadogMetric := range ds.store {
+		if filter(datadogMetric) {
+			datadogMetrics = append(datadogMetrics, datadogMetric)
+		}
+	}
+
+	return datadogMetrics
+}
+
+// Count returns number of elements in store
+func (ds *DatadogMetricsInternalStore) Count() int {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return len(ds.store)
+}
+
+// Set `DatadogMetricInternal` for id
+func (ds *DatadogMetricsInternalStore) Set(id string, datadogMetric model.DatadogMetricInternal, sender string) {
+	ds.lock.Lock()
+	ds.store[id] = datadogMetric
+	ds.lock.Unlock()
+
+	ds.notify(setOperation, id, sender)
+}
+
+// Delete `DatadogMetricInternal` corresponding to id if present
+func (ds *DatadogMetricsInternalStore) Delete(id, sender string) {
+	ds.lock.Lock()
+	_, exists := ds.store[id]
+	delete(ds.store, id)
+	ds.lock.Unlock()
+
+	if exists {
+		ds.notify(deleteOperation, id, sender)
+	}
+}
+
+// LockRead allows to get an item and leave the store in a locked state to allow safe Read -> Operation -> Write sequences
+// Still locks if the key does not exist as you may want to prevent a concurrent Write.
+// It's not very efficient to lock the whole store but it's probably enough for our use case.
+func (ds *DatadogMetricsInternalStore) LockRead(id string, lockOnMissing bool) *model.DatadogMetricInternal {
```

Comment:
Yes, it would. The only other option to make sure we don't loose any write is to introduce a `Version` that can be compared and implementing a retry everywhere we do a store (which may not be trivial), it's better in terms of locking but adds significant complexity.